src/effects/native/echoeffect.cpp

```diff
@@ -143,90 +144,98 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
         } else if (period < 1/8.0) {
             period = 1/8.0;
         }
-        delay_samples = period * groupFeatures.beat_length_sec
-                * sampleRate * EchoGroupState::kChannelCount;
+        delay_frames = period * groupFeatures.beat_length_sec * sampleRate;
     } else {
         // period is a number of seconds
         period = std::max(period, 1/8.0);
-        delay_samples = period * sampleRate * EchoGroupState::kChannelCount;
+        delay_frames = period * sampleRate;
     }
-    VERIFY_OR_DEBUG_ASSERT(delay_samples > 0) {
-        delay_samples = 1;
+    VERIFY_OR_DEBUG_ASSERT(delay_frames > 0) {
+        delay_frames = 1;
     }
+
+    int delay_samples = delay_frames * gs.kChannelCount;
     VERIFY_OR_DEBUG_ASSERT(delay_samples <= gs.delay_buf.size()) {
         delay_samples = gs.delay_buf.size();
     }
 
-    if (period < gs.prev_period) {
-        // If the delay time has shrunk, we may need to wrap the write position.
-        gs.write_position = gs.write_position % delay_samples;
-    } else if (period > gs.prev_period) {
-        // If the delay time has grown, we need to zero out the new portion
-        // of the buffer we are using.
-        SampleUtil::applyGain(gs.delay_buf.data(gs.prev_delay_samples), 0,
-                              gs.delay_buf.size() - gs.prev_delay_samples);
-    }
-
+    int prev_read_position = gs.write_position;
+    DECREMENT_RING(prev_read_position, gs.prev_delay_samples, gs.delay_buf.size());
     int read_position = gs.write_position;
+    DECREMENT_RING(read_position, delay_samples, gs.delay_buf.size());
 
     // Feedback the delay buffer and then add the new input.
     const CSAMPLE_GAIN send_delta = (send_amount - gs.prev_send) /
             (numSamples / EchoGroupState::kChannelCount);
-    const CSAMPLE_GAIN send_start = send_amount + send_delta;
+    const CSAMPLE_GAIN send_start = gs.prev_send + send_delta;
+
+    const CSAMPLE_GAIN feedback_delta = (feedback_amount - gs.prev_feedback) /
+            (numSamples / EchoGroupState::kChannelCount);
+    const CSAMPLE_GAIN feedback_start = gs.prev_feedback + feedback_delta;
+
     for (unsigned int i = 0; i < numSamples; i += EchoGroupState::kChannelCount) {
-        CSAMPLE_GAIN send_ramped = send_start;
-        if (send_delta > 0.0) {
-            send_ramped += send_delta * i / EchoGroupState::kChannelCount;
+        CSAMPLE_GAIN send_ramped = send_start
+                + send_delta * i / EchoGroupState::kChannelCount;
+        CSAMPLE_GAIN feedback_ramped = feedback_start
+                + feedback_delta * i / EchoGroupState::kChannelCount;
+
+        CSAMPLE bufferedSample0 = gs.delay_buf[read_position];
+        CSAMPLE bufferedSample1 = gs.delay_buf[read_position + 1];
```

Comment:
I think using "Left" and "Right" in the variable names would be more readable than "0" and "1".